### PR TITLE
Add rate-limiters to ServerPolicy

### PR DESCRIPTION
**Pull request by alpeb**
This adds the local_rate_limit module to the server-policy crate, that
`ServerPolicy` uses for its new `local_rate_limit` field, containing
three optional rate-limiters: total, identity, overrides (this one is
really a vector of limiters, one per configured override).

I tried putting that under `Protocol` instead, but the `PartialEq`
requirement made it very hard to follow. `Server` OTOH doesn't really
require that trait, so I was able to remove it and accommodate the
limiters.

I made sure to avoid pulling the dashmap dependency in `governor`; ~I
haven't checked yet the necessity of the "jitter" and "quanta" features.~

This temporarily overrides linkerd2-proxy-api dependency to pick changes
from linkerd/linkerd2-proxy-api#388

## Update

The `HttpPolicyService` middleware has been expanded to call the `local_rate_limit` module above to perform the rate-limit check.

### CODE REVIEW

File: linkerd/proxy/server-policy/Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ publish = false
 proto = ["linkerd-http-route/proto", "linkerd2-proxy-api", "prost-types"]
 
 [dependencies]
+governor = { version = "0.6", default-features = false, features = ["std", "jitter", "quanta"] }
```
Comment (olix0r) [inline linkerd/proxy/server-policy/Cargo.toml:13]:
```suggestion
governor = { version = "0.6", default-features = false, features = ["std", "jitter"] }
```

We're not using the quanta clock so it's fine to omit it.

File: linkerd/proxy/server-policy/src/local_rate_limit.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    pub total: Option<RateLimiter<NotKeyed, InMemoryState, DefaultClock>>,
+    pub identity: Option<RateLimiter<String, DefaultKeyedStateStore<String>, DefaultClock>>,
+    pub overrides: Vec<HttpLocalRateLimitOverride>,
+}
+
+#[derive(Debug)]
+pub struct HttpLocalRateLimitOverride {
+    pub ids: Vec<String>,
+    pub rate_limit: RateLimiter<Vec<String>, DefaultKeyedStateStore<Vec<String>>, DefaultClock>,
+}
```
Comment (olix0r) [inline linkerd/proxy/server-policy/src/local_rate_limit.rs:19]: * I think we can maintain a narrower API--no need to expose all of the governor types, we
* It would be preferable to use the Identity types directly here rather than require String round-tripping (allocations).
* I think the override rate limiter is misconfigured in your version--it should not be keyed on Vec<String>.
* Furthermore, it seems preferable to store the overrides for efficient lookup (i.e. in a HashMap).

```suggestion
#[derive(Debug, Default)]
pub struct LocalRateLimit {
    total: Option<RateLimit<InMemoryState>>,
    per_identity: Option<RateLimit<DefaultKeyedStateStore<Option<Id>>>>,
    overrides: HashMap<Id, Arc<RateLimit<InMemoryState>>>,
}

#[derive(Debug)]
struct RateLimit<S: StateStore> {
    rps: NonZeroU32,
    limiter: RateLimiter<S::Key, S, DefaultClock>,
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum RateLimitError {
    #[error("total rate limit exceeded: {0}rps")]
    Total(NonZeroU32),
    #[error("per-identity rate limit exceeded: {0}rps")]
    PerIdentity(NonZeroU32),
    #[error("override rate limit exceeded: {0}rps")]
    Override(NonZeroU32),
}
```

```rust
impl LocalRateLimit {
    pub fn check(&self, id: Option<&Id>) -> Result<(), RateLimitError> {
        if let Some(lib) = &self.total {
            if lim.limiter.check().is_err() {
                return Err(RateLimitError::Total(lim.rps));
            }
        }

        if let Some(id) = id {
            if let Some(lim) = self.overrides.get(id) {
                if lim.limiter.check().is_err() {
                    return Err(RateLimitError::Override(lim.rps));
                }
                return Ok(());
            }
        }

        if let Some(lim) = &self.per_identity {
            if limiter.check_key(&lim.id.cloned()).is_err() {
                return Err(RateLimitError::PerIdentity(lim.rps));
            }
        }

        Ok(())
    }
}
```

The server-policy crate can hide all of the specifics about rate limiting and let callers have a very simple interface.

File: linkerd/proxy/server-policy/src/local_rate_limit.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                            .into_iter()
+                            .flat_map(|cl| cl.identities.into_iter().map(|id| id.name))
+                            .collect();
+                        let quota =
+                            Quota::per_second(NonZeroU32::new(lim.requests_per_second).unwrap());
+                        let rate_limit = RateLimiter::keyed(quota);
+                        HttpLocalRateLimitOverride { ids, rate_limit }
+                    })
+                })
+                .collect();
```
Comment (olix0r) [inline linkerd/proxy/server-policy/src/local_rate_limit.rs:63]: Note that we cannot call `unwrap` (or expect) when handling input from the control plane. We need to make sure the None-ness propagates to cause the item to be ignored:
```suggestion
         let total = proto.total.and_then(|lim| {
                let rps = NonZeroU32::new(lim.requests_per_second)?;
                let limiter = RateLimiter::direct(Quota::per_second(rps));
                Some(RateLimit { rps, limiter })
            });

            let per_identity = proto.identity.and_then(|lim| {
                let rps = NonZeroU32::new(lim.requests_per_second)?;
                let limiter = RateLimiter::hashmap(Quota::per_second(rps));
                Some(RateLimit { rps, limiter })
            });

            let overrides = proto
                .overrides
                .into_iter()
                .flat_map(|ovr| {
                    let Some(lim) = ovr.limit else {
                        return vec![];
                    };

                    let Some(rps) = NonZeroU32::new(lim.requests_per_second) else {
                        return vec![];
                    };
                    let limiter = RateLimiter::direct(Quota::per_second(rps));
                    let limit = Arc::new(RateLimit { rps, limiter });

                    ovr.clients
                        .into_iter()
                        .flat_map(|cl| {
                            cl.identities
                                .into_iter()
                                .filter_map(|id| id.name.parse::<Id>().ok())
                        })
                        .map(move |id| (id, limit.clone()))
                        .collect::<Vec<(Id, Arc<RateLimit<InMemoryState>>)>>()
                })
                .collect::<HashMap<Id, Arc<RateLimit<InMemoryState>>>>();

```

File: linkerd/proxy/server-policy/src/lib.rs [hunk 4]
```diff
@@ -14,10 +16,11 @@ pub use self::{
 };
 pub use linkerd_http_route as route;
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug)]
 pub struct ServerPolicy {
     pub protocol: Protocol,
     pub meta: Arc<Meta>,
+    pub local_rate_limit: Arc<HttpLocalRateLimit>,
```
Comment (olix0r) [inline linkerd/proxy/server-policy/src/lib.rs:23]: This isn't a big problem, but it seems like it would be most consistent (with both the proxy prior art and API) for the rate limit to be encoded in the `Protocol`, since it's not relevant for certain protocol variants.

```rust
    Detect {
        http: Arc<[http::Route]>,
        http_local_rate_limit: Arc<LocalRateLimit>
        timeout: time::Duration,
        tcp_authorizations: Arc<[Authorization]>,
    },
    Http1(Arc<[http::Route]>, Arc<LocalRateLimit>),
    Http2(Arc<[http::Route]>, Arc<LocalRateLimit>),
```

File: linkerd/proxy/server-policy/src/local_rate_limit.rs (hunk 2 above)
Comment (alpeb) [reply linkerd/proxy/server-policy/src/local_rate_limit.rs:19]: > I think the override rate limiter is misconfigured in your version--it should not be keyed on Vec.

As per linkerd/linkerd2#13231, an `HTTPLocalRateLimitPolicy` would look something like this:
```yaml
apiVersion: policy.linkerd.io/v1alpha1
kind: HTTPLocalRateLimitPolicy
metadata:
  namespace: emojivoto
  name: web-rl
spec:
  targetRef:
    group: policy.linkerd.io
    kind: Server
    name: web-http
  total:
    requestsPerSecond: 100
  identity:
    requestsPerSecond: 20
  overrides:
  - requestsPerSecond: 10
    clientRefs:
    - kind: ServiceAccount
      namespace: emojivoto
      name: default
```
Under overrides, I implemented the `clientRefs` as an array, which explains the types I was hinting at (I may have made that an array _after_ our conversations, sorry about that). It's definitely simpler to have it be just one object though like in your suggestion, right?

File: linkerd/proxy/server-policy/src/lib.rs (hunk 4 above)
Comment (alpeb) [reply linkerd/proxy/server-policy/src/lib.rs:23]: The problem is that governor's `RateLimiter` doesn't implement `PartialEq`, which `Protocol` requires. That requirement can be lifted from the types depending on `Protocol`, except for `ServerPermit`, which uses it for test assertions. So we could either manually implement `PartialEq` for `ServerPermit` just so that those tests aren't affected, or implement it for our `LocalRateLimit` wrapper you suggested above but it'd only check for equality on rps and ids. Any thoughts?

File: linkerd/proxy/server-policy/src/local_rate_limit.rs (hunk 2 above)
Comment (olix0r) [reply linkerd/proxy/server-policy/src/local_rate_limit.rs:19]: The override representation in the YAML doesn't dictate how we store them at runtime.

Also note that I interpret the overrides to support "pooled" limits. For example:

```yaml
  - requestsPerSecond: 10
    clientRefs:
    - kind: ServiceAccount
      namespace: emojivoto
      name: web
    clientRefs:
    - kind: ServiceAccount
      namespace: emojivoto
      name: default
```

In this situation I would expect these two accounts to _share_ a rate limit of 10RPS. This is distinct from the configuration in which they have distinct but identical limits:

```yaml
  - requestsPerSecond: 10
    clientRefs:
    - kind: ServiceAccount
      namespace: emojivoto
      name: web
  - requestsPerSecond: 10
    clientRefs:
    - kind: ServiceAccount
      namespace: emojivoto
      name: default
```

I imagine that we could build a map of overrides from the existing spec:

```rust

            let overrides = proto
                .overrides
                .into_iter()
                .flat_map(|ovr| {
                    let Some(lim) = ovr.limit else {
                        return vec![];
                    };

                    let Some(rps) = NonZeroU32::new(lim.requests_per_second) else {
                        return vec![];
                    };
                    let limiter = RateLimiter::direct(Quota::per_second(rps));
                    let limit = Arc::new(RateLimit { rps, limiter });

                    ovr.clients
                        .into_iter()
                        .flat_map(|cl| {
                            cl.identities
                                .into_iter()
                                .filter_map(|id| id.name.parse::<Id>().ok())
                        })
                        .map(move |id| (id, limit.clone()))
                        .collect::<Vec<(Id, Arc<RateLimit<InMemoryState>>)>>()
                })
                .collect::<HashMap<Id, Arc<RateLimit<InMemoryState>>>>();
```

This supports shared limits and efficient map-based lookups.

File: linkerd/proxy/server-policy/src/lib.rs (hunk 4 above)
Comment (olix0r) [reply linkerd/proxy/server-policy/src/lib.rs:23]: After looking at your follow-on PR, this change probably isn't worth it -- it's more complicated to consume the rate limit without much benefit.

File: linkerd/proxy/server-policy/src/lib.rs [hunk 5]
```diff
@@ -131,6 +135,28 @@ pub mod proto {
...
+                        local_rate_limit,
+                        ..
+                    })
+                    | api::proxy_protocol::Kind::Http2(api::proxy_protocol::Http2 {
+                        local_rate_limit,
+                        ..
+                    }) => local_rate_limit.unwrap_or_default().into(),
+                    _ => Default::default(),
+                }
+            };
```
Comment (olix0r) [inline linkerd/proxy/server-policy/src/lib.rs:158]: nit: We can make this slightly more compact by removing redundant braces. The other blocks that do this define temporary variables.

```suggestion
            let local_rate_limit = match protocol
                .clone()
                .and_then(|api::ProxyProtocol { kind }| kind)
                .ok_or(InvalidServer::MissingProxyProtocol)?
            {
                api::proxy_protocol::Kind::Detect(api::proxy_protocol::Detect {
                    http_local_rate_limit,
                    ..
                }) => http_local_rate_limit.unwrap_or_default().into(),
                api::proxy_protocol::Kind::Http1(api::proxy_protocol::Http1 {
                    local_rate_limit,
                    ..
                })
                | api::proxy_protocol::Kind::Http2(api::proxy_protocol::Http2 {
                    local_rate_limit,
                    ..
                }) => local_rate_limit.unwrap_or_default().into(),
                _ => Default::default(),
            };
```

File: linkerd/app/inbound/src/policy/http.rs [hunk 6]
```diff
@@ -171,6 +171,8 @@ where
             }
         };
 
+        try_fut!(self.check_rate_limit());
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http.rs:174]: It would be good to exercise this in a [test](https://github.com/linkerd/linkerd2-proxy/blob/b580e657e565d77e4138c34397fea7182fd1797b/linkerd/app/inbound/src/policy/http/tests.rs).

File: linkerd/proxy/server-policy/src/local_rate_limit.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    #[error("total rate limit exceeded: {0}rps")]
+    Total(NonZeroU32),
+    #[error("per-identity rate limit exceeded: {0}rps")]
+    PerIdentity(NonZeroU32),
+    #[error("override rate limit exceeded: {0}rps")]
+    Override(NonZeroU32),
+}
+
+impl LocalRateLimit {
+    pub fn check(&self, id: Option<&Id>) -> Result<(), RateLimitError> {
```
[the remark this answers is not in the document]
Comment (alpeb) [reply linkerd/proxy/server-policy/src/local_rate_limit.rs:30]: Sounds good :+1:  I will be testing using governor's [FakeRelativeClock](https://docs.rs/governor/latest/governor/clock/struct.FakeRelativeClock.html) for this, given they don't rely on tokio.

File: linkerd/proxy/server-policy/src/local_rate_limit.rs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    #[error("per-identity rate limit exceeded: {0}rps")]
+    PerIdentity(NonZeroU32),
+    #[error("override rate limit exceeded: {0}rps")]
+    Override(NonZeroU32),
+}
+
+// === impl LocalRateLimit ===
+
+impl LocalRateLimit {
+    pub fn new_no_overrides(
```
Comment (olix0r) [inline linkerd/proxy/server-policy/src/local_rate_limit.rs:48]: We would typically [feature-gate this](https://github.com/linkerd/linkerd2-proxy/pull/3328) to assert that it's only used for tests.

File: linkerd/app/inbound/src/policy/http/tests.rs [hunk 9]
```diff
@@ -363,6 +375,38 @@ async fn http_filter_inject_failure() {
...
+                .body(hyper::Body::default())
+                .unwrap(),
+        )
+        .await
+        .expect("serves");
+    let permit = rsp
+        .extensions()
+        .get::<HttpRoutePermit>()
+        .expect("permitted");
+    assert_eq!(permit.labels.route.route, rmeta);
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http/tests.rs:407]: Is there some way that we could actually test actually fails when a rate limit is exceeded? E.g. Set of rate limit of 1 and send 2 requests.

File: linkerd/app/inbound/src/policy/http/tests.rs [hunk 10]
```diff
@@ -1,6 +1,7 @@
 use super::*;
 use crate::policy::{Authentication, Authorization, Meta, Protocol, ServerPolicy};
 use linkerd_app_core::{svc::Service, Infallible};
+use linkerd_proxy_server_policy::local_rate_limit::LocalRateLimit;
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http/tests.rs:4]: With a reexport we can simplify this:
```suggestion
use linkerd_proxy_server_policy::LocalRateLimit;
```

File: linkerd/app/inbound/src/http/server.rs [hunk 11]
```diff
@@ -203,6 +203,12 @@ impl errors::HttpRescue<Error> for ServerRescue {
             ));
         }
 
+        if errors::is_caused_by::<linkerd_proxy_server_policy::local_rate_limit::RateLimitError>(
```
Comment (olix0r) [inline linkerd/app/inbound/src/http/server.rs:206]:
```suggestion
        if errors::is_caused_by::<linkerd_proxy_server_policy::RateLimitError>(
```

File: linkerd/app/inbound/src/policy/http.rs [hunk 12]
```diff
@@ -287,6 +289,21 @@ impl<T, N> HttpPolicyService<T, N> {
             .route_not_found(labels, self.connection.dst, self.connection.tls.clone());
         HttpRouteNotFound(()).into()
     }
+
+    fn check_rate_limit(&self) -> Result<()> {
+        let id = match self.connection.tls {
+            Conditional::Some(tls::ServerTls::Established {
```
[the remark this answers is not in the document]
Comment (alpeb) [reply linkerd/app/inbound/src/policy/http.rs:295]: That's actually not reexported in the tls module. Should I add it?

File: linkerd/app/inbound/src/policy/http/tests.rs (hunk 9 above)
Comment (alpeb) [reply linkerd/app/inbound/src/policy/http/tests.rs:411]: Yep, just added in my latest commit :+1:

File: linkerd/app/inbound/src/policy/http.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/app/inbound/src/policy/http.rs:295]: Ah I was thinking of:
```suggestion
            tls::ConditionalServerTls::Some(tls::ServerTls::Established {
```

I think it's fine to leave as-is with the Conditional import.

File: linkerd/app/inbound/src/policy/http/tests.rs [hunk 13]
```diff
@@ -363,6 +375,110 @@ async fn http_filter_inject_failure() {
...
+                .body(hyper::Body::default())
+                .unwrap(),
+        )
+        .await
+        .expect("serves");
+    let permit = rsp
+        .extensions()
+        .get::<HttpRoutePermit>()
+        .expect("permitted");
+    assert_eq!(permit.labels.route.route, rmeta);
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http/tests.rs:411]: We can omit the checks that are covered by other tests--they're not necessary to test the rate limiting behavior.

```suggestion
    assert_eq!(rsp.status(), ::http::StatusCode::OK);
```

File: linkerd/app/inbound/src/policy/http/tests.rs [hunk 14]
```diff
@@ -363,6 +375,110 @@ async fn http_filter_inject_failure() {
...
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http/tests.rs:426]:
```suggestion
    assert_eq!(rsp.status(), ::http::StatusCode::OK);
```

File: linkerd/app/inbound/src/policy/http/tests.rs [hunk 15]
```diff
@@ -363,6 +375,110 @@ async fn http_filter_inject_failure() {
...
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http/tests.rs:462]:
```suggestion
    assert_eq!(rsp.status(), ::http::StatusCode::OK);
```

File: linkerd/app/inbound/src/policy/http/tests.rs [hunk 16]
```diff
@@ -363,6 +375,110 @@ async fn http_filter_inject_failure() {
...
+        )
+        .await
+        .expect_err("should deny");
+    let err = rsp
+        .downcast_ref::<RateLimitError>()
+        .expect("rate limit error");
+    match err {
+        RateLimitError::PerIdentity(rps) => assert_eq!(rps, &std::num::NonZeroU32::new(1).unwrap()),
+        _ => panic!("unexpected error"),
+    };
```
Comment (olix0r) [inline linkerd/app/inbound/src/policy/http/tests.rs:479]: nit: slightly more readable:

```suggestion
    match err {
        RateLimitError::PerIdentity(rps) => assert_eq!(rps.into(), 1),
        _ => panic!("unexpected error"),
    }
```

**Review by olix0r — APPROVED** [review]
Take note that we'll want to update the commit message to reflect the current state of this branch.

File: linkerd/app/inbound/src/policy/http/tests.rs (hunk 16 above)
Comment (alpeb) [reply linkerd/app/inbound/src/policy/http/tests.rs:479]: Clippy doesn't like that :disappointed: 
```
error[E0283]: type annotations needed
   --> linkerd/app/inbound/src/policy/http/tests.rs:465:60
    |
465 |         RateLimitError::PerIdentity(rps) => assert_eq!(rps.into(), 1),
    |                                                            ^^^^
    |
    = note: cannot satisfy `_: std::convert::From<&std::num::NonZeroU32>`
    = note: required for `&std::num::NonZeroU32` to implement `std::convert::Into<_>`
help: try using a fully qualified path to specify the expected types
    |
465 |         RateLimitError::PerIdentity(rps) => assert_eq!(<&std::num::NonZeroU32 as std::convert::Into<T>>::into(rps), 1),
    |                                                        +++++++++++++++++++++++++++++++++++++++++++++++++++++++   ~
```
